File: VietOCR/tags/Release-1.9/src/net/sourceforge/vietocr/SpellChecker.java
/**
 * Copyright @ 2010 Quan Nguyen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.sourceforge.vietocr;

import com.stibocatalog.hunspell.Hunspell;
import java.awt.Color;
//import java.util.logging.*;
import java.io.*;
import java.text.BreakIterator;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;
import java.util.regex.*;
import javax.swing.JOptionPane;
import javax.swing.event.DocumentEvent;
import javax.swing.event.DocumentListener;
import javax.swing.text.*;

public class SpellChecker {

    JTextComponent textComp;
    // define the highlighter
    Highlighter.HighlightPainter myPainter = new WavyLineHighlighter(Color.red);
    String localeId;
    File baseDir;
    static Properties prop;
    static List<DocumentListener> lstList = new ArrayList<DocumentListener>();
    Hunspell.Dictionary spellDict;
    static List<String> userWordList = new ArrayList<String>();
    static long mapLastModified = Long.MIN_VALUE;

    public SpellChecker(JTextComponent textComp, String langCode) {
        this.textComp = textComp;
        baseDir = Utilities.getBaseDir(SpellChecker.this);

        File xmlFile = new File(baseDir, "data/ISO639-1.xml");

        try {
            if (prop == null) {
                prop = new Properties();
                prop.loadFromXML(new FileInputStream(xmlFile));
            }
            localeId = prop.getProperty(langCode);
            if (localeId == null) {
                localeId = prop.getProperty(langCode.substring(0, 3));
                if (localeId == null) {
                    JOptionPane.showMessageDialog(null, "Need to add an entry in data/ISO639-1.xml file.", Gui.APP_NAME, JOptionPane.ERROR_MESSAGE);
                }
            }
        } catch (IOException ioe) {
            JOptionPane.showMessageDialog(null, "Missing ISO639-1.xml file. Cannot find it in " + new File(baseDir, "data").getPath() + " directory.", Gui.APP_NAME, JOptionPane.ERROR_MESSAGE);
        }
    }

    public void enableSpellCheck() {
        if (localeId == null) {
            return;
        }
        try {
            spellDict = Hunspell.getInstance().getDictionary(new File(baseDir, "dict/" + localeId).getPath());
            loadUserDictionary();

            SpellcheckDocumentListener docListener = new SpellcheckDocumentListener();
            lstList.add(docListener);
            this.textComp.getDocument().addDocumentListener(docListener);
            spellCheck();
        } catch (Exception e) {
            JOptionPane.showMessageDialog(null, e.getMessage(), Gui.APP_NAME, JOptionPane.ERROR_MESSAGE);
        }
    }

    void spellCheck() {
        List<String> words = parseText(textComp.getText());
        List<String> misspelledWords = spellCheck(words);
        if (misspelledWords.isEmpty()) {
            return; // perfect writer!
        }

        StringBuilder sb = new StringBuilder();
        for (String word : misspelledWords) {
            sb.append(word).append("|");
        }
        sb.setLength(sb.length() - 1); //remove last |

        // build regex
        String patternStr = "\\b(" + sb.toString() + ")\\b";

        Pattern pattern = Pattern.compile(patternStr, Pattern.CASE_INSENSITIVE);
        Matcher matcher = pattern.matcher(textComp.getText());

        Highlighter hi = textComp.getHighlighter();
        hi.removeAllHighlights();

        while (matcher.find()) {
            try {
                hi.addHighlight(matcher.start(), matcher.end(), myPainter);
            } catch (BadLocationException ex) {
//                Logger.getLogger(Gui.class.getName()).log(Level.SEVERE, null, ex);
            }
        }
    }

    List<String> spellCheck(List<String> words) {
        List<String> misspelled = new ArrayList<String>();
        for (String word : words) {
            if (spellDict.misspelled(word)) {
                // is mispelled word in user.dic?
                if (!userWordList.contains(word.toLowerCase())) {
                    misspelled.add(word);
                }
            }
        }

        return misspelled;
    }

    List<String> parseText(String text) {
        List<String> words = new ArrayList<String>();
        BreakIterator boundary = BreakIterator.getWordInstance();
        boundary.setText(text);
        int start = boundary.first();
        for (int end = boundary.next(); end != BreakIterator.DONE; start = end, end = boundary.next()) {
            if (!Character.isLetter(text.charAt(start))) {
                continue;
            }
            words.add(text.substring(start, end));
        }

        return words;
    }

    public void disableSpellCheck() {
        if (localeId == null) {
            return;
        }
        this.textComp.getDocument().removeDocumentListener(lstList.remove(0));
        this.textComp.getHighlighter().removeAllHighlights();
    }

    /**
     * Loads user dictionary only once.
     */
    void loadUserDictionary() {
        try {
            File userDict = new File(baseDir, "dict/user.dic");
            long fileLastModified = userDict.lastModified();

            if (fileLastModified <= mapLastModified) {
                return;// no need to reload dictionary
            }

            mapLastModified = fileLastModified;
            userWordList.clear();
            BufferedReader in = new BufferedReader(new InputStreamReader(new FileInputStream(userDict), "UTF8"));
            String str;
            while ((str = in.readLine()) != null) {
                userWordList.add(str.toLowerCase());
            }
            in.close();
        } catch (IOException e) {
            JOptionPane.showMessageDialog(null, "Cannot find \"user.dic\" in " + new File(baseDir, "dict").getPath() + " directory.", Gui.APP_NAME, JOptionPane.ERROR_MESSAGE);
        }
    }

    class SpellcheckDocumentListener implements DocumentListener {

        @Override
        public void removeUpdate(DocumentEvent e) {
            spellCheck();
        }

        @Override
        public void insertUpdate(DocumentEvent e) {
            spellCheck();
        }

        @Override
        public void changedUpdate(DocumentEvent e) {
        }
    }
}

File: VietOCR.NET/tags/Release-1.9/SpellChecker.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Xml;
using System.IO;
using System.Windows.Forms;
using System.Drawing;
using NHunspell;
using System.Text.RegularExpressions;
using System.Collections;

namespace VietOCR.NET
{
    class SpellChecker
    {
        TextBoxBase textbox;
        string localeId;
        String workingDir;
        static ArrayList listeners = new ArrayList();
        static List<CharacterRange> spellingErrorRanges = new List<CharacterRange>();
        static List<String> userWordList = new List<String>();
        static DateTime mapLastModified = DateTime.MinValue;
        Hunspell spellDict;
        CustomPaintTextBox cntl;

        public CharacterRange[] GetSpellingErrorRanges()
        {
            return spellingErrorRanges.ToArray();
        }


        public SpellChecker(TextBoxBase textbox, string langCode)
        {
            this.textbox = textbox;

            XmlDocument doc = new XmlDocument();

            workingDir = Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location);
            String xmlFilePath = Path.Combine(workingDir, "Data/ISO639-1.xml");
            Dictionary<string, string> ht = new Dictionary<string, string>();
            doc.Load(xmlFilePath);

            XmlNodeList list = doc.GetElementsByTagName("entry");
            foreach (XmlNode node in list)
            {
                ht.Add(node.Attributes[0].Value, node.InnerText);
            }

            localeId = ht[langCode.Substring(0, 3)];
        }

        public void enableSpellCheck()
        {
            if (localeId == null)
            {
                return;
            }
            try
            {
                string dictPath = workingDir + "/dict/" + localeId;
                spellDict = new Hunspell(dictPath + ".aff", dictPath + ".dic");
                LoadUserDictionary();

                listeners.Add(new System.EventHandler(this.textbox_TextChanged));

                this.textbox.TextChanged += (System.EventHandler)listeners[0];
                cntl = new CustomPaintTextBox(this.textbox, this);

                spellCheck();
            }
            catch (Exception e)
            {
                MessageBox.Show(e.Message);
            }
        }

        void spellCheck()
        {
            spellingErrorRanges.Clear();
            List<String> words = parseText(textbox.Text);
            List<String> misspelledWords = spellCheck(words);
            if (misspelledWords.Count == 0)
            {
                return;
            }


            StringBuilder sb = new StringBuilder();
            foreach (String word in misspelledWords)
            {
                sb.Append(word).Append("|");
            }
            sb.Length -= 1; //remove last |

            // build regex
            String patternStr = "\\b(" + sb.ToString() + ")\\b";
            Regex regex = new Regex(patternStr, RegexOptions.IgnoreCase);
            MatchCollection mc = regex.Matches(textbox.Text);

            // Loop through  the match collection to retrieve all 
            // matches and positions.
            for (int i = 0; i < mc.Count; i++)
            {
                spellingErrorRanges.Add(new CharacterRange(mc[i].Index, mc[i].Length));
            }
            //new CustomPaintTextBox(textbox, this);
        }

        List<String> spellCheck(List<String> words)
        {
            List<String> misspelled = new List<String>();

            foreach (String word in words)
            {
                if (!spellDict.Spell(word))
                {
                    // is mispelled word in user.dic?
                    if (!userWordList.Contains(word.ToLower()))
                    {
                        misspelled.Add(word);
                    }

                }
            }

            return misspelled;
        }

        List<String> parseText(String text)
        {
            List<String> words = new List<String>();
            BreakIterator boundary = BreakIterator.GetWordInstance();
            boundary.Text = text;
            int start = boundary.First();
            for (int end = boundary.Next(); end != BreakIterator.DONE; start = end, end = boundary.Next())
            {
                if (!Char.IsLetter(text[start]))
                {
                    continue;
                }
                words.Add(text.Substring(start, end - start));
            }

            return words;
        }

        public void disableSpellCheck()
        {
            if (localeId == null)
            {
                return;
            }

            if (listeners.Count > 0)
            {
                this.textbox.TextChanged -= (System.EventHandler)listeners[0];
                listeners.RemoveAt(0);
            }
            spellingErrorRanges.Clear();
        }

        private void textbox_TextChanged(object sender, EventArgs e)
        {
            spellCheck();
            this.textbox.Invalidate();
        }

        void LoadUserDictionary()
        {
            try
            {
                String baseDir = Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location);
                string strUserDictFile = Path.Combine(baseDir, @"dict\user.dic");
                FileInfo userDict = new FileInfo(strUserDictFile);
                DateTime fileLastModified = userDict.LastWriteTime;

                if (fileLastModified <= mapLastModified)
                {
                    return; // no need to reload dictionary
                }

                mapLastModified = fileLastModified;
                userWordList.Clear();

                using (StreamReader sr = new StreamReader(strUserDictFile, Encoding.UTF8))
                {
                    string str;

                    while ((str = sr.ReadLine()) != null)
                    {
                        userWordList.Add(str.ToLower());
                    }
                    //sr.Close();
                }
            }
            catch (IOException e)
            {
                MessageBox.Show(e.Message);
            }
        }
    }
}
